### ISSUE TITLE
Compilation is slower than in previous versions

### ISSUE THREAD

**Issue by fxgallego** [dolphinsmalltalk/Dolphin#932]
In our build process we install ~ 800 packages.
This are the times for 2 different releases:
7.1.6 (bc89c5b93a15a3599eb6599287445f4649554e3a) -> 6m:02s
7.1.11 (89549ae2fdfc41c5dfed4bf3323a4de891b2f3b3) ->  6m:39s

I didn't any research to detect where the problem is, if it's in the compiler side or in the image side.
This is something that doesn't affect us too much, only in build times due to the amount of source code we load, but I wanted to report it because maybe something can be done to improve this.

If there are suspicions that some modification could cause this, I can help testing.

**Comment by fxgallego** [comment]
I've tried 7.1.10 and got times equal to 7.1.6 so it should be after that.
I'll try to do the binary search trough the commits. Not sure when, but due to the lockdown I have extra time 😄. I'll close it if I give up.

**Comment by blairmcg** [comment]
Well it seems that you've done the first pivot and eliminated 6 months of the period already 🙂.

I should have mentioned that 7.1.10 is immediately before the change to the compiler for #854. I tagged a version beforehand because it was a risky change. You might want to try the commit immediately after 7.1.10 (1f60d3b9ba860d95dd988c6ff03a21cd5412cbd0) to eliminate (or blame) that.

**Comment by fxgallego** [comment]
Differences in time seems to be in some way equally distributed across different commits after 7.1.10 so I don't believe this is going to produce any interesting findings. Maybe I could identify each suspicious commit but that's is difficult because my measurements are not exact and they could drive to false positives.

This make my think: how much slower would be a Smalltalk version of the compiler?

**Comment by blairmcg** [comment]
That's a shame. I would have expected that there might be a few percent difference from some of the C primitives, but not necessarily that it would have affected the compiler that much. 

BTW: If you normally run your Windows 10 machine in the default "Balanced" power mode, I would recommend switching it to "High Performance" mode if you want to run benchmarks. Aside from Dolphin running quite a lot faster, the benchmark times will be more consistent from run to run.

The compiler written in C is not very efficient because of the way it is designed. It doesn't build an AST, or an intermediate representation, but rather attempts to compiler straight to bytecode, which it then spends a lot of time doing repeated micro optimisation passes on. Also when it was originally written Dolphin didn't have proper block closures with clean, copying and full block variants. The analysis required for that had to be tacked on and it made more complicated even that it is fundamentally by the lack of intermediate representations. Even using a language that compiles to fast code won't help you that much if the design is wrong. Also the callbacks it has to make into Smalltalk are relatively costly. In other words, I wouldn't assume it will be massively slower in Smalltalk, although it is definitely going to be slower.

We can run an experiment that demonstrates this reasonably well. It is generally understood that the lexical analysis phase is the most expensive compilation phase just simply because it processes the input character by character, and everything else is working with orders of magnitude less data. A reasonable approximation might be that lexical analysis is about 40% of the time for a bytecode compiler that doesn't do a lot of complicated optimisation at code generation time. 

We can easily measure the time it takes for the Smalltalk parser to build parse trees for every method (i.e. to lexically _and_ syntactically analyse every method):

```smalltalk
[Class allMethodsDo: [:each | each parseTree analyze]] timeToRun humanReadablePrintString "7.4 s"
```
Now compare that with actually re-compiling using the C-based compiler:
```smalltalk
[Object recompileAll] timeToRun humanReadablePrintString "5.54 s"
```
So, yes, it will be slower, but not by the 10-15x that you might expect from the relative execution speed.

**Comment by fxgallego** [comment]
Thank for the explanation. I'll try to give this some time in the weekend.

**Comment by fxgallego** [comment]
@blairmcg I tried `[Object recompileAll] timeToRun humanReadablePrintString` in my image, after load all my packages and it only takes  ~ 30s to compile all methods. That's really impressive, my changes is 114 MB.
And that's interesting because installing all packages is about 6 minutes and doesn't includes the Dolphin packages. I know it's not the same, but something that point me to profile `PackageManager>>install:` with a package that I previously removed.

![image](https://user-images.githubusercontent.com/2080171/84548383-dcc5fe80-acdb-11ea-82cc-b70c09662d6e.png)

`Package declareClasses` loads all classes declared to avoid class pre-requisite problems.
After that `ChunkSourceFiler fileIn` evaluates chunks for the classes and because now they are modifications instead of new classses, the class builder runs a class migration and that's time consuming. Seems to me that maybe doing the things in a different way could bring massive improvements in the file-in and in consequence this compiler increase of 10% would be unnoticeable.

**Comment by blairmcg** [comment]
I certainly agree with the sentiment. The best solution to most performance issues is to do less, rather than do the same thing faster.

The `ClassBuilder` should not need to do a mutation for classes where the definition has not changed, and I think this is working. However, it will be doing an initial mutation of the forward reference declaration. I don't see such stark figures as you when I try profiling (for example) the load of the base tests. However this may be because your image is much larger. You might also not be running with the parallel `oneWayBecome:` if you are not building the VM yourself, as this is not in the released builds. Since there are two of those per `ClassBuilder` mutation, it will make a difference. The more objects you have, the more the gain. I 

The class builder mutations do perform memory scans, as well as a lot of other stuff. In fact they do 4 - an instance scan and a one-way become for the metaclass and the class. The instance scan for the metaclass instance is unnecessary since there is a single instance, and we know what it is. I just tried a small fix for that, and it made no significant difference for me in loading the base tests package on my own machine. That isn't entirely surprising as the instance scans are also performed in parallel and are sub-millisecond on my machine with a standard image. It may be more important in your larger image though.

I also noticed that the `declareClasses` operation spends (relative to other stuff it does anyway), a lot of time sorting the subclass collections. There are a number of possible ways to improve that, but rather than change too much at once I just made another small and simple change to speed up the sort order comparison for `Metaclass`. That did make a measurable difference in overall load time for the base tests package (which adds a lot of subclasses to some specific classes), but it caused some unexpected breaks so I am still investigating that.

**Comment by fxgallego** [comment]
I saw the sorting taking some time in the profiler, but in my case, it has less incidence than the mutations and sure that's because the image size. Anyway, if we can avoid one of them, the other will start to show as important.

I've commented out `declareClasses` in `loadPAX` and the load times  where cut by more than a half. In a small package running with `declareClasses` -> 716 ms and without running it -> 309 ms.

Why is there `declareClasses`? If the classes are defined after in the package source, and they are in the right order, why I need it?
For example, this small package loads without problems, or at least it seems to load well 😄

**Comment by blairmcg** [comment]
That is a very good question. From the comments in the code this appears to be because the source globals in the package are loaded before the classes are otherwise defined and might refer to the classes. That order seems a bit questionable at first glance, but the source globals (globals defined as source code expressions) are generally pool dictionaries, the members of which may be referred to from code. It seems a bit of a corner case to have source globals that refer to classes in the package but are still loadable before those classes, but presumably there may have been some examples at some point that motivated this as it seems like a non-obvious choice.

We could be radical and remove it altogether, and in many (most, perhaps all existing) cases that will work. A less radical thing might be to make it an option (and turn on the option for all older packages so they behave as before). A completely safe thing to do (and therefore the thing I will try now) would be to skip it if there are no source globals - the majority of packages don't have them anyway.

**Comment by blairmcg** [comment]
Ah, and there is also the "loose" (extension) methods. These are loaded before the classes in the package. It isn't a corner case that these might refer to classes. Again the load order is probably wrong here. Although these are in the pax file itself, they should be loaded after the classes. Changing the load order is a potentially breaking change, however, because of ##() expressions. It is a bit obscure and unreliable to depend on being able to execute code from the previously loaded loose methods when loading the classes, however. Class initialize methods are not a problem as these are all run together after all the code has been loaded. 

Before changing the load order we would have to try it out in practice, and it would need bake time. I don't think that should be done in a minor release, but of course you could always apply a patch in the meantime.

**Comment by fxgallego** [comment]
FYI I've loaded all my code with `declareClasses` commented out without a glitch.
Time with `declareClassses` -> 6m34s
Time without `declareClasses` -> 2m43s
This made my day and it also contributes to reduce my carbon footprint !😄

**Comment by blairmcg** [comment]
That's good. My concern about the loose methods is unfounded. Now that I have bothered to check, I can see that these follow the class definitions in the .pax, so the only reason would be to support loading the source globals.

It should be possible to detect the rare corner cases where this is needed and then mark the package with a property for those cases. However I'm not going to do the analysis to see if that is needed now, but will put that on the backlog for 7.2.

So I'll go back to my plan of making the safe change of just skipping the declaration if there are no source globals. This should cover the majority of cases anyway.

**Comment by fxgallego** [comment]
I think this can be closed for the following reasons:
1. The compiler is fast enough.
2. The VM code is changing in a fast pace so minor differences are expected, some looses, some gains.
3. A compiler back end in Smalltalk is in the backlog #119
4. A huge improvement in package install was achieved in c84b8da76f5d449551cfdee514dc324383b57955 and previous commits.

**Comment by rko281** [comment]
I'm seeing a warning in the latest image loading some packages (Seaside, Willow) which use Pharo-style shared pools. I'm filing this as a comment here rather than a new bug report since I think it represents an undocumented feature in Dolphin(?). Example classes:
```
Object subclass: #MyPool
	instanceVariableNames: ''
	classVariableNames: 'PoolVar1 PoolVar2 PoolVar3'
	poolDictionaries: ''
	classInstanceVariableNames: ''

Object subclass: #MyClass
	instanceVariableNames: ''
	classVariableNames: ''
	poolDictionaries: 'MyPool'
	classInstanceVariableNames: ''
```
Provided I implement:
```
associationsDo: aBlock

	^self classPool associationsDo: aBlock
```
...as a class method in MyPool to keep the autocomplete happy (I normally do this by having the pool class as a subclass of SharedPool from my Portishead repository) I can refer to MyPool's class variables (PoolVar1 etc.) in MyClass in the same way I would "normal" pool variables. This matches Pharo's pools mechanism and so makes ports more straightforward. 

If I load a package containing this arrangement in the latest D7.1 image I get the warning
```
Pool dictionary #MyPool does not exist.
Proceed to create it as an empty constants pool.

Click OK to bypass this warning and continue with the installation, or Cancel to abort. 
[OK] [Cancel]
```
Hitting OK allows the package to load, however I didn't get this message prior to the changes discussed here (slight lie - previously I saw this same warning when loading a .PAC package, but not when using .PAX; I normally use the latter).

As I say, I'm not sure this constitutes a bug as I'm not sure this is officially supported behaviour, however I found this a useful feature which it would be nice to have working again.

[Test package here](https://www.dropbox.com/s/xm6maflbhsgbxy1/class%20pool%20test.7z?dl=1)

**Comment by blairmcg** [comment]
Yes, I can understand that that wouldn't work as it requires that the classes that are representing the pools be at least declared before they are referenced, which they would have been before but not now. A cheesy workaround in the short term would be to put an empty source global (like a dummy pool constants dictionary) into the package.

I was thinking of adding a property to the package to force forward declaration, and then relying on that being manually set for the cases where it is actually required and removing the dynamic check on the presence of source globals. This could be defaulted to true for backwards compatibility now, and changed to a default of false in 7.2.

I had also been thinking of adopting the Pharo representation for pools too. It has certain advantages as there is no longer a separate representation, and the initializer code can be anything and directly edited in the image.

**Comment by rko281** [comment]
Thanks; I've done my own cheesy workaround for now of manually editing the .pax file so the pool class appears before its users. The offending packages are fairly static so this is fine for the moment.
